adds documentation generation using doxygen

cc10512 opened the pull request:
Documentation generation relies on the ax_prog_doxygen macros. By
default we generate HTML and PDF documentation which includes class
hierarchies generated using the 'dot' tool. The macros provide
enable/disable flags for different features of doxygen (see configure
--help).

The configuration for doxygen is in the docs/doxygen directory. The
p4c_layout.xml file is used to customize the look of the HTML
documentation (this is still work in progress, as the default layout
is not very friendly).

Documentation is produced from source files (.def, .h, .cpp) and from
markdown (.md) files.  To ensure that introductory chapters are
produced in a pre-defined order, markdown files in the docs/doxygen
directory follow the convention xx_name.md, where xx is the ordinal
number of the chapter.

Still TODO:
* integrate documentation for the extension directories
* define chapters for documenting components of the compiler and hook
  them up in the navigation layout.
* more customization of the HTML documentation
* figure out how to rename the section for all the internal README.md
  files

Fixes #383

sethfowler commented:
Doxygen does often end up being useless but it can be very useful when used pervasively in a project. Try browsing around in [Clang's Doxygen documentation](https://clang.llvm.org/doxygen/modules.html).

When I was writing libclang bindings for other languages I found that documentation invaluable. It'd be great if we get to the point where this project has a similar level of documentation.

cc10512 commented:
@mbudiu-vmw: I have disabled the documentation building -- needs to be explicitly enabled using --enable-doxygen-doc to configure. 
I also added the dependencies to the README files and a section on how to write documentation.
I also added headings to the README.md files that get pulled in, so now the documentation looks much better. Please give it a try :-)

cc10512 commented:
@ChrisDodd: yes, based on @mbudiu-vmw request, it is disabled by default in bootstrap.sh. Run with --enable-doxygen-doc to enable it.
